### PR TITLE
store/tikv: speed up buildTask().

### PR DESCRIPTION
**Pull request by disksing**
@shenli @coocood @zimulala

### CODE REVIEW

**Comment by disksing** [comment]
Result of "TestBuildHugeTasks":
Before:
2016/09/01 16:49:02 coprocessor.go:142: [warning] buildCopTasks takes too much time (12.371190552s), range len 26000000, task len 26

After:
2016/09/01 16:43:04 coprocessor.go:265: [warning] buildCopTasks takes too much time (748.301153ms), range len 26000000, task len 26

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -107,35 +109,161 @@ const (
...
+		}
+		// All rest ranges belong to the same region.
+		if i == ranges.len() {
+			appendTask(region, ranges)
+			break
+		}
+
+		r := ranges.at(i)
+		if region.Contains(r.StartKey) {
+			// Part of r is not in the region. We need to split it.
```
Comment (shenli) [inline store/tikv/coprocessor.go]: if bytes.Equal(region.EndKey(), r.EndKey) && region.Contains(r.StartKey), the whole r is in region.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@zimulala @tiancaiamao PTAL

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -107,35 +109,161 @@ const (
...
+		tasks = append(tasks, &copTask{
+			idx:      len(tasks),
+			region:   region,
+			status:   taskNew,
+			ranges:   ranges,
+			respChan: make(chan *coprocessor.Response, 1),
+		})
+	}
+
+	for ranges.len() > 0 {
```
Comment (zimulala) [inline store/tikv/coprocessor.go]: ranges.len() -> rangesLen ?

File: store/tikv/coprocessor.go [hunk 3]
```diff
@@ -107,35 +109,161 @@ const (
...
+		// Iterate to the first range that is not complete in the region.
+		var i int
+		for ; i < ranges.len(); i++ {
+			r := ranges.at(i)
+			if !(region.Contains(r.EndKey) || bytes.Equal(region.EndKey(), r.EndKey)) {
+				break
+			}
+		}
+		// All rest ranges belong to the same region.
+		if i == ranges.len() {
```
Comment (zimulala) [inline store/tikv/coprocessor.go]: The same as line 219.

File: store/tikv/coprocessor.go [hunk 4]
```diff
@@ -107,35 +109,161 @@ const (
...
 	return ranges
 }
 
-func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges []kv.KeyRange, desc bool) ([]*copTask, error) {
+func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, desc bool) ([]*copTask, error) {
+	start := time.Now()
+	rangesLen := ranges.len()
+
 	var tasks []*copTask
-	for _, r := range ranges {
```
Comment (tiancaiamao) [inline store/tikv/coprocessor.go]: this change looks complex...
closure is also bad for performance

what about simply change line 129 to `tasks := make([]*copTask, 0, len(ranges))`

Comment (disksing) [inline store/tikv/coprocessor.go]: Task length does not always equal to ranges (some ranges may belong to the same region).

**Comment by zimulala** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**